### PR TITLE
Throw TranslogCorruptedException in more cases

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today we do not throw a `TranslogCorruptedException` in certain cases of
translog corruption, such as for a corrupted checkpoint file or when an
expected file (either checkpoint or translog) is completely missing. This means
that `elasticsearch-shard` will not truncate the translog in those cases.

This commit strengthens the translog corruption tests to corrupt and/or delete
both translog and checkpoint files, and ensures that a
`TranslogCorruptedException` is thrown in all cases. It also sometimes
simulates a recovery after a crash while rolling the translog generation,
including cases where the rolled checkpoint contains incorrect data.

It also adjusts (and renames) `RemoveCorruptedShardDataCommandIT.getDirs()` to
return only a single path, since in practice this was the only thing that could
happen and yet we were relying on its callers to verify this and not all
callers were doing so.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/index/shard/RemoveCorruptedShardDataCommandIT.java [hunk 1]
```diff
@@ -339,30 +337,28 @@ public void testCorruptTranslogTruncation() throws Exception {
                 @Override
                 public Settings onNodeStopped(String nodeName) throws Exception {
                     // and we can actually truncate the translog
-                    for (Path translogDir : translogDirs) {
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/index/shard/RemoveCorruptedShardDataCommandIT.java]: I'd look at this bit ignoring whitespace changes if I were you :)

File: server/src/test/java/org/elasticsearch/index/shard/RemoveCorruptedShardDataCommandIT.java [hunk 2]
```diff
@@ -504,15 +500,13 @@ public void testCorruptTranslogTruncationOfReplica() throws Exception {
         final MockTerminal terminal = new MockTerminal();
         final OptionParser parser = command.getParser();
 
-        for (Path translogDir : translogDirs) {
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/index/shard/RemoveCorruptedShardDataCommandIT.java]: Mostly whitespace changes here.

**Comment by DaveCTurner** [comment]
@elasticmachine please run elasticsearch-ci/2

File: server/src/test/java/org/elasticsearch/index/translog/TestTranslog.java [hunk 3]
```diff
@@ -56,41 +58,94 @@
...
         logger.info("--> corruptRandomTranslogFile: translogDir [{}], minUsedTranslogGen [{}]", translogDir, minGeneration);
+
+        Path unnecessaryCheckpointCopyPath = null;
+        try {
+            final Path checkpointPath = translogDir.resolve(CHECKPOINT_FILE_NAME);
+            final Checkpoint checkpoint = Checkpoint.read(checkpointPath);
+            unnecessaryCheckpointCopyPath = translogDir.resolve(Translog.getCommitCheckpointFileName(checkpoint.generation));
+            if (LuceneTestCase.rarely(random) && Files.exists(unnecessaryCheckpointCopyPath) == false) {
+                // if we crashed while rolling a generation then we might have copied `translog.ckp` to its numbered generation file but
+                // have not yet written a new `translog.ckp`. During recovery we must also verify that this file is intact, so it's ok to
```
Comment (dnhatn) [inline server/src/test/java/org/elasticsearch/index/translog/TestTranslog.java]: ++

**Review by dnhatn — COMMENTED** [review]
This looks good. I left some comment. Thanks @DaveCTurner.

File: server/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 4]
```diff
@@ -255,8 +254,9 @@ public Translog(
             if (Files.exists(commitCheckpoint)) {
                 Checkpoint checkpointFromDisk = Checkpoint.read(commitCheckpoint);
                 if (checkpoint.equals(checkpointFromDisk) == false) {
-                    throw new IllegalStateException("Checkpoint file " + commitCheckpoint.getFileName() +
-                        " already exists but has corrupted content expected: " + checkpoint + " but got: " + checkpointFromDisk);
+                    throw new TranslogCorruptedException(commitCheckpoint.toString(),
+                        new IllegalStateException("Checkpoint file " + commitCheckpoint.getFileName() +
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/index/translog/Translog.java]: Thanks, see 333a189ef353ed62455eaeea46e78f13d9c7e5b7.

File: server/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 5]
```diff
@@ -219,28 +219,27 @@ public Translog(
...
                 if (Files.exists(committedTranslogFile) == false) {
-                    throw new IllegalStateException("translog file doesn't exist with generation: " + i + " recovering from: " +
-                        minGenerationToRecoverFrom + " checkpoint: " + checkpoint.generation + " - translog ids must be consecutive");
+                    throw new TranslogCorruptedException(committedTranslogFile.toString(), new IllegalStateException(
+                        "translog file doesn't exist with generation: " + i + " recovering from: " +
+                        minGenerationToRecoverFrom + " checkpoint: " + checkpoint.generation + " - translog ids must be consecutive"));
                 }
                 final TranslogReader reader = openReader(committedTranslogFile,
-                    Checkpoint.read(location.resolve(getCommitCheckpointFileName(i))));
+                    i == checkpoint.generation ? checkpoint : Checkpoint.read(location.resolve(getCommitCheckpointFileName(i))));
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/index/translog/Translog.java]: Sure, see abef96a48c2219748b1902db8f50e2ad9cb6aad4.

File: server/src/test/java/org/elasticsearch/index/shard/RemoveCorruptedShardDataCommandIT.java [hunk 6]
```diff
@@ -314,60 +310,64 @@ public void testCorruptTranslogTruncation() throws Exception {
...
-                    .setIndex(indexName).setShard(0).setPrimary(true)
-                    .get().getExplanation();
-
-            final UnassignedInfo unassignedInfo = explanation.getUnassignedInfo();
+            final UnassignedInfo unassignedInfo = client().admin().cluster().prepareAllocationExplain()
+                .setIndex(indexName).setShard(0).setPrimary(true).get().getExplanation().getUnassignedInfo();
             assertThat(unassignedInfo.getReason(), equalTo(UnassignedInfo.Reason.ALLOCATION_FAILED));
+            Throwable throwable = unassignedInfo.getFailure();
+            while (throwable != null) {
+                if (throwable instanceof TranslogCorruptedException) {
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/index/shard/RemoveCorruptedShardDataCommandIT.java]: We can indeed. 77a521af87c96a10807247bd9eeeee768d1f930f.

File: server/src/test/java/org/elasticsearch/index/store/CorruptedTranslogIT.java [hunk 7]
```diff
@@ -19,127 +19,92 @@
...
-            if (Files.isDirectory(translogDir)) {
-                translogDirs.add(translogDir);
+        assertBusy(() -> {
+            final ClusterAllocationExplainResponse allocationExplainResponse
+                = client().admin().cluster().prepareAllocationExplain().setIndex("test").setShard(0).setPrimary(true).get();
+            final UnassignedInfo unassignedInfo = allocationExplainResponse.getExplanation().getUnassignedInfo();
+            assertThat(unassignedInfo, not(nullValue()));
+            final Exception outerException = unassignedInfo.getFailure();
+            Throwable throwable = outerException;
+            while (throwable != null) {
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/index/store/CorruptedTranslogIT.java]: We can indeed: 77a521af87c96a10807247bd9eeeee768d1f930f

**Review by DaveCTurner — COMMENTED** [review]
Thanks @dnhatn, everything actioned.

File: libs/core/src/main/java/org/elasticsearch/core/internal/io/IOUtils.java [hunk 8]
```diff
@@ -186,15 +186,11 @@ public static void deleteFilesIgnoringExceptions(final Collection<? extends Path
...
         if (!unremoved.isEmpty()) {
-            final StringBuilder b = new StringBuilder("could not remove the following files (in the order of attempts):\n");
+            final IOException ioException = new IOException("could not remove all files");
             for (final Map.Entry<Path,Throwable> kv : unremoved.entrySet()) {
-                b.append("   ")
-                        .append(kv.getKey().toAbsolutePath())
-                        .append(": ")
-                        .append(kv.getValue())
-                        .append("\n");
+                ioException.addSuppressed(new IOException(kv.getKey().toAbsolutePath().toString(), kv.getValue()));
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply libs/core/src/main/java/org/elasticsearch/core/internal/io/IOUtils.java]: Meh the tests for `IOUtils` don't let me make this change, and I don't want to chase that up in this PR, so I've reverted it.

**Review by dnhatn — APPROVED** [review]
LGTM

**Comment by csoulios** [comment]
@DaveCTurner can we merge this PR  to 6.x branch?

**Pull request merged**